test/spec/ol/source/tileimage.test.js

```diff
@@ -172,4 +173,67 @@ describe('ol.source.TileImage', function() {
       tile.load();
     });
   });
+
+  describe('tile load events', function() {
+
+    var source;
+
+    beforeEach(function() {
+      source = new ol.source.TileImage({
+        url: '{z}/{x}/{y}'
+      });
+    });
+
+    it('dispatches tileloadstart and tileloadend events', function() {
+      source.setTileLoadFunction(function(tile) {
+        tile.setState(ol.TileState.LOADED);
+      });
+      var startSpy = sinon.spy();
+      source.on('tileloadstart', startSpy);
+      var endSpy = sinon.spy();
+      source.on('tileloadend', endSpy);
+      var tile = source.getTile(0, 0, -1, 1, ol.proj.get('EPSG:3857'));
+      tile.load();
+      expect(startSpy.callCount).to.be(1);
+      expect(endSpy.callCount).to.be(1);
+    });
+
+    it('dispatches tileerror events', function(done) {
+      source.setTileLoadFunction(function(tile) {
+        tile.setState(
+            tile.state == ol.TileState.ERROR ? ol.TileState.LOADED : ol.TileState.ERROR);
+      });
+      var startSpy = sinon.spy();
+      source.on('tileloadstart', startSpy);
+      var errorSpy = sinon.spy();
+      source.on('tileloaderror', function(e) {
+        setTimeout(function() {
+          e.tile.setState(ol.TileState.LOADING);
+          e.tile.setState(ol.TileState.LOADED);
+        }, 0);
+        errorSpy();
+      });
+      source.on('tileloadend', function() {
+        expect(startSpy.callCount).to.be(2);
+        expect(errorSpy.callCount).to.be(1);
```

Comment:
This takes a while to understand.  And I'm still not sure I get it.  It looks like the tile state is set to `LOADED` twice in this test (once at 204 and once at 212).  But `tileloadend` only gets called once?  Or will it get called a second time but we are already `done()`?

I wonder if there is a simpler test to confirm that `ERROR` state results in both `tileloaderror` and `tileloadend` being fired (if that is the correct behavior).

Reply:
Does this pass?

```js
source.setTileLoadFunction(function(tile) {
  tile.setState(ol.TileState.ERROR);
});

var errorSpy = sinon.spy();
source.on('tileloaderror', errorSpy);
source.on('tileloadend', function() {
  expect(errorSpy.callCount).to.be(1);
  done();
});
```

Reply:
Actually, maybe that isn't what this is about.

Reply:
The life cycle of the tile is `IDLE` - `LOADING` - `ERROR` - `LOADING` - `LOADED`.

I think this is reasonable, because users might respond to `ERROR` by loading the tile again, finally resulting in a `LOADED` state. L211-L212 mimics what a user would do to reload the tile. In a real application, the `LOADING` state would be set again by calling [`ol.ImageTile#load()`](https://github.com/openlayers/openlayers/blob/2a117cce8f08f9f31306fa4d5ae5d645b1508aa3/src/ol/imagetile.js#L125-L137), which sets up `load` and `error` listeners for an `Image`.

Reply:
Got it.  I started with the test description (it "dispatches tileerror events") and was just surprised/confused by how involved the sequence was to assert that.